authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,998 @@
+            change.getSecurableObject().fullName());
+        // If it exists policy with the same privilege, then directly return true, because support
+        // idempotent operation.
+        return true;
+      }
+    } else {
+      policy = new RangerPolicy();
+      policy.setService(rangerServiceName);
+      policy.setName(change.getSecurableObject().fullName());
+      policy.setPolicyLabels(Lists.newArrayList(MANAGED_BY_GRAVITINO));
+
+      List<String> nsMetadataObject =
+          Lists.newArrayList(
+              SecurableObjects.DOT_SPLITTER.splitToList(change.getSecurableObject().fullName()));
+      if (nsMetadataObject.size() > 4) {
+        // The max level of the securable object is `catalog.db.table.column`
+        throw new RuntimeException("The securable object than 4");
+      }
+      nsMetadataObject.remove(0); // remove `catalog`
+
+      List<String> rangerDefinesList =
+          Lists.newArrayList(
+              RangerDefines.RESOURCE_DATABASE,
+              RangerDefines.RESOURCE_TABLE,
+              RangerDefines.RESOURCE_COLUMN);
+      for (int i = 0; i < nsMetadataObject.size(); i++) {
+        RangerPolicy.RangerPolicyResource policyResource =
+            new RangerPolicy.RangerPolicyResource(nsMetadataObject.get(i));
+        policy.getResources().put(rangerDefinesList.get(i), policyResource);
+      }
+    }
+
+    addPolicyItem(policy, change.getRoleName(), change.getSecurableObject());
+    try {
+      if (policy.getId() == null) {
+        rangerClient.createPolicy(policy);
+      } else {
+        rangerClient.updatePolicy(policy.getId(), policy);
+      }
+    } catch (RangerServiceException e) {
+      throw new RuntimeException(e);
+    }
+
+    return true;
+  }
+
+  /**
+   * Remove the securable object's privilege from the policy. <br>
+   * Because Ranger uses unique metadata to location Ranger policy and manages all the privileges,
+   * So A Ranger policy maybe contains multiple Gravitino privilege objects. <br>
+   * Remove Ranger policy item condition is: <br>
+   * 1. This Ranger policy item's accesses are equal the Gravitino securable object's privilege.
+   * <br>
+   * 2. This Ranger policy item's users and groups are empty. <br>
+   * If policy didn't have any policy item, then delete this policy: <br>
+   */
+  private boolean doRemoveSecurableObject(
+      String roleName, RoleChange.RemoveSecurableObject change) {
+    RangerPolicy policy = findManagedPolicy(change.getSecurableObject());
+
+    if (policy == null) {
+      LOG.warn(
+          "Cannot find the Ranger policy({}) for the Gravitino securable object({})!",
+          roleName,
+          change.getSecurableObject().fullName());
+      // Don't throw exception or return false, because need support immutable operation.
+      return true;
+    }
+
+    List<RangerPolicyItem> policyItems = policy.getPolicyItems();
+    policyItems.forEach(
+        policyItem -> {
+          boolean match =
```

Comment:
Maybe we can extract this to a method to let other update/remove/add securable object method to reuse it?